blaze/expr/collections.py

```diff
@@ -418,6 +418,7 @@ def on_right(self):
         return on_right
 
     @property
+    @memoize
     def schema(self):
```

Comment:
> adding __weakref__ to Join.__slots__ led to this cryptic error:

Why not just add `__slots__ = ()` to `Node` and `Expr`? That'll prevent that error from occurring.